### PR TITLE
fix: deletion logic for transformed checks

### PR DESCRIPTION
**Pull request by yashmehrotra**
Fixes: https://mission-control.azure.lab.flanksource.com/incidents/0188ba77-539b-f928-9abe-75b84caf97dc and https://mission-control.azure.lab.flanksource.com/incidents/018929f8-be24-2368-7983-a4e47955d5b5

### CODE REVIEW

File: pkg/db/canary.go [hunk 1]
```diff
@@ -117,20 +135,21 @@ func GetTransformedCheckIDs(canaryID string) ([]string, error) {
...
-		Error
+
+	// Alertmanager checks are marked as healthy on deletion
+	query := `
+        UPDATE checks
+        SET deleted_at = NOW(),
+            status = (CASE WHEN checks.type = 'alertmanager' THEN 'healthy'
+                           ELSE checks.status
+                      END)
+        WHERE id IN (?)
```
Comment (yashmehrotra) [inline pkg/db/canary.go:150]: @moshloop Moving the transformed checks' status setting logic at the query level

File: pkg/db/canary.go [hunk 2]
```diff
@@ -121,22 +139,43 @@ func GetTransformedCheckIDs(canaryID string) ([]string, error) {
...
+	updates := map[string]any{
+		"deleted_at": gorm.Expr("NOW()"),
+	}
+	if status != "" {
+		if !utils.Contains(models.CheckHealthStatuses, status) {
+			return fmt.Errorf("invalid check health status: %s", status)
+		}
+		updates["status"] = status
 	}
 	return Gorm.Table("checks").
```
Comment (moshloop) [inline pkg/db/canary.go:155]: I think we want to insert a new check rather than update the old checks

File: pkg/jobs/canary/canary_jobs.go [hunk 3]
```diff
@@ -66,54 +65,75 @@ var minimumTimeBetweenCanaryRuns = 10 * time.Second
...
-	checksToMarkHealthy := utils.SetDifference(existingTransformedChecks, newChecksCreated)
-	if err := db.UpdateChecksStatus(checksToMarkHealthy, models.CheckStatusHealthy); err != nil {
-		logger.Errorf("error deleting transformed checks for canary %s: %v", job.Canary.GetPersistedID(), err)
+	checkDeleteStrategyGroup := make(map[string][]string)
+	checksToRemove := utils.SetDifference(existingTransformedChecks, transformedChecksCreated)
+	if len(checksToRemove) > 0 && len(transformedChecksCreated) > 0 {
+		for _, checkID := range checksToRemove {
+			strategy := checkIDDeleteStrategyMap[checkID]
+			var status string
+			if strategy == "MarkHealthy" {
```
Comment (moshloop) [inline pkg/jobs/canary/canary_jobs.go:121]: Can we use Constants ?
I think ignore should be the default strategy - i.e. do not insert or update anything, just mark as deleted.

File: pkg/db/canary.go (hunk 2 above)
Comment (yashmehrotra) [reply pkg/db/canary.go:155]: Do you mean check_status ?
Ie. Insert new check status and mark the existing check as deleted ?

File: pkg/jobs/canary/canary_jobs.go (hunk 3 above)
Comment (yashmehrotra) [reply pkg/jobs/canary/canary_jobs.go:121]: 👍🏽

**Pull request merged**